### PR TITLE
Quality-of-life API improvements.

### LINKED ISSUE #31: Conversion between data structures and BitVec
I'm using vers-vecs for a project, and it's great, but one part of the project is shaping up to really, _really_ want to be implemented in terms of taking a constructed tree and replacing a node with another tree, or replacing the children of a node with some other trees (this is looking down the line, so I'm not sure of the exact needs. Some cases I can work around, but some are looking like they really need to do this kind of stuff).  Is this something you'd be open to considering a PR for?

I'm happy to write up a PR with a suggested implementation and such, but before getting into the weeds of that I thought some discussion would be good, since I assume you'll have opinions on the API, or maybe even don't want the functionality at all in the library.

An alternative which would make it possible to for users of the library to implement this kind of stuff on their own would be to make it possible to convert the succinct datatypes "backwards". For example for BpTree if it had a method like:

```rs
fn into_rsvec(self) -> RsVec {
    self.vec
}
```

and probably similar for RsVec -> BitVec, it'd be possible to compute the needed operations on the succinct structures, decay them to simpler datatypes, do the surgery required, and rebuild the succinct structures.

### PR DESCRIPTION
**Pull request by arnsholt**
As we discussed in #31, From<> implementations for the types that seem to make sense (for EliasFano `.iter().collect()` should suffice, and the RMQ types are Deref and aren't constructed from BitVecs). While using the library I've also felt the need for the extend and split functionality, so I added those while I was at it.

Adds:
- `From<T> for BitVec` implementations for converting RsVec, BpTree and WaveletMatrix back into BitVecs
- `impl Extend<BitVec> for BitVec` and `BitVec.extend_vec()` to append many bits at once
- `BitVec.split_at()` and `._split_at_unchecked()` to split a BitVec into two halves

### CODE REVIEW

**Comment by arnsholt** [comment]
And I see now that between me looking at this again yesterday and deciding to do it, and actually doing it today, you did most of it anyways. Oops. =)

**Comment by Cydhra** [comment]
Well it doesn't matter, too much, the main work I did yesterday was fix the padding thing in b478799e3a26a6a33c815fc5210316bd1e090831, which enables the `From` implementation to work in the first place. The work on the `issue_31` branch isn't very substantial. You could rebase onto the branch though, if you want, as I already did the documentation work for the conversion methods, so you could rewire your `From` implementations to call those methods and profit from the existing documentation. If you think this is redundant, we can also ignore the branch.

File: src/wavelet/mod.rs [hunk 1]
```diff
@@ -2053,6 +2053,14 @@ impl WaveletMatrix {
     }
 }
 
+impl From<WaveletMatrix> for BitVec {
```
Comment (Cydhra) [inline src/wavelet/mod.rs]: I don't understand the significance of this. A wavelet matrix encodes symbols bit per bit, the n'th bit of each symbol in the n'th RsVec of the WaveletMatrix. Why would we want to append all those vectors into a single vector?

File: src/bit_vec/mod.rs [hunk 2]
```diff
@@ -736,6 +736,19 @@ impl BitVec {
         self.len += len;
     }
 
+    /// Append the bits of another bit vector to the vector.
+    pub fn extend_bitvec(&mut self, other: &Self) {
```
Comment (Cydhra) [inline src/bit_vec/mod.rs:743]: We need test cases for all the methods. I can implement them if you enabled the option that I can edit the PR.

**Review by Cydhra — CHANGES_REQUESTED** [review]

File: src/wavelet/mod.rs (hunk 1 above)
Comment (arnsholt) [reply src/wavelet/mod.rs]: It's not something I've needed in my project, I added mostly for symmetry with BpTree. Since both are constructed from a BitVec, I figured it made sense to implement the backwards conversion for both types. If that's nonsense for WaveletMatrix we can delete it.

**Comment by arnsholt** [comment]
I rebased onto your branch and updated the From impls for RsVec and BpTree. Left WaveletMatrix as is for now.

I'm pretty sure I added edit for you on the PR, so if you have a clear idea of what tests you want you can put them in. Otherwise, I can try to add some tests tomorrow.

File: src/wavelet/mod.rs (hunk 1 above)
Comment (Cydhra) [reply src/wavelet/mod.rs]: Yeah it doesn't make sense here. The wavelet matrix is constructed from integers, or for very strange inputs from BitVecs, but they are transposed during construction, so appending them like this doesn't make sense.

**Review by Cydhra — APPROVED** [review]
I have added test cases and fixed three bugs in split_at. I also removed the `From<WaveletMatrix>` implementation.

File: src/bit_vec/mod.rs [hunk 3]
```diff
@@ -1194,6 +1215,71 @@ impl BitVec {
...
+        let full_limbs = self.len / WORD_SIZE;
+
+        // if we start in the middle of a limb, we need to copy the leading partial limb.
+        // however, we limit the range to the size of the other vector, since we could be starting
+        // in the last limb of the original vector
+        let leading_partial = min((WORD_SIZE - (at % WORD_SIZE)) % WORD_SIZE, other_len);
+
+        let iter_limbs = if leading_partial > 0 {
+            other.append_bits_unchecked(
+                self.data[first_limb] >> (WORD_SIZE - leading_partial),
```
Comment (Cydhra) [inline src/bit_vec/mod.rs]: This fails if the limb isn't `WORD_SIZE` long, see red test case.

**Comment by arnsholt** [comment]
Wow, I clearly didn't think the splitting logic through properly. Sorry for rushing making the PR and making you find the bugs first! I had a proper think about the algorithm and ended up with a factoring where the initial logic was more or less sound, but with an added edge case that needed handling where the split point is inside a partial last limb of the input vector. I also added a few more test cases.

**Comment by arnsholt** [comment]
And I completely failed to see the unchecked set of tests just above. Sorry for being so sloppy again!

**Comment by arnsholt** [comment]
There, now it passes the full test suite here.
